python/dgl/core.py

```diff
@@ -228,13 +228,20 @@ def data_dict_to_list(graph, data_dict, func, target):
     if isinstance(func, fn.BinaryMessageFunction):
         if target in ["u", "v"]:
             output_list = [None] * graph._graph.number_of_ntypes()
-            for srctype, _, dsttype in graph.canonical_etypes:
+            if not isinstance(data_dict, dict):
```

Comment:
are we assuming `data_dict` is either `defaultdict` or `dict`?

Reply:
We assume here data_dict is a dict or a tensor.